File: SDAC/Esercitazioni/Esercitazione 4/c-source/sort.c
#include <stdio.h>

#include "sort.h"

//------------------------MERGE SORT------------------------------
void merge(int arr[], int left, int mid, int right) {
    int i, j, k;        int n1 = mid-left+1;    int n2 = right-mid;

    int L[n1], R[n2];       //Creazione di array temporanei
    for(i = 0; i < n1; ++i) //Assegnazione dei valori da sinistra
        L[i] = arr[left+i];
    for(j = 0; j < n2; ++j) //Assegnazione dei valori dal centro a destra
        R[j] = arr[mid+1+j];

    i = 0; j = 0; 
    k = left;
    while(i < n1 && j < n2) { //inserisco tutti i 
         if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    // Copia gli elementi rimanenti di L[], se ce ne sono
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    // Copia gli elementi rimanenti di R[], se ce ne sono
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

void mergeSort_aus(int arr[], int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort_aus(arr, left, mid);
        mergeSort_aus(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

void mergeSort(array* a) {
    mergeSort_aus(a-> arr, 0, (a-> size)-1);
    return;
}

//---------------------------HEAP SORT------------------------------
void maxHeap(int arr[], int n, int i) {
    int largest = i;    int left = 2*i +1;      int right = 2*i + 2;

    // Se il figlio sinistro è più grande della radice
    if (left < n && arr[left] > arr[largest])
        largest = left;
    // Se il figlio destro è più grande della radice più grande finora
    if (right < n && arr[right] > arr[largest])
        largest = right;
    // Se il più grande non è la radice
    if (largest != i) {
        swap(&arr[i], &arr[largest]);
        // Ricorsivamente esegue il maxHeap sul sottoalbero interessato
        maxHeap(arr, n, largest);
    }
}

void heapSort(array *a) {
    int n = a-> size;
    int* arr = a-> arr;
    
    // Costruisci un heap max
    for (int i = n / 2 - 1; i >= 0; i--)
        maxHeap(arr, n, i);

    // Estrai elementi uno per uno dall'heap
    for (int i = n - 1; i > 0; i--) {
        // Sposta la radice (l'elemento più grande) alla fine
        swap(&arr[0], &arr[i]);
        // Chiamata ricorsiva maxHeapify sul ridotto heap
        maxHeap(arr, i, 0);
    }
    return;
}

//---------------------INSERTION SORT-------------------------------
void insertionSortRicorsivo(int arr[], int n) {
    if(n <= 1) return;

    //Ordino ricorsivamente il sottoaray escludendo l'ultimo elemento
    insertionSortRicorsivo(arr, n-1);

    int last = arr[n-1];
    int j = n-2;

    //Sposta gli elemento maggiori di last verso destra
    while(j >= 0 && arr[j] > last) {
        arr[j+1] = arr[j];
        j--;
    }
    arr[j+1] = last;
}

void insertionSort(array *a) {
    insertionSortRicorsivo(a-> arr, a-> size);
    return;
}

//------------------------SELECTION SORT------------------------------
void selectionSortRecursive(int arr[], int n, int index) {
    // Caso base: se l'indice è uguale alla dimensione dell'array, l'ordinamento è completo
    if (index == n - 1)
        return;

    // Trova l'indice del minimo elemento nel sottoarray non ordinato
    int minIndex = index;
    for (int i = index + 1; i < n; i++) {
        if (arr[i] < arr[minIndex])
            minIndex = i;
    }

    // Scambia l'elemento minimo con l'elemento corrente
    swap(&arr[index], &arr[minIndex]);

    for(int i = 0; i < n; ++i) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    // Ordina ricorsivamente il sottoarray escludendo l'elemento appena ordinato
    selectionSortRecursive(arr, n, index + 1);
}

void selectionSort(array *a) {
    selectionSortRecursive(a-> arr, a-> size, 0);

    return;
}

//--------------------------QUICK SORT------------------------------
int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j <= high - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return i + 1;
}

void quickSortRecursive(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);

        quickSortRecursive(arr, low, pi - 1);
        quickSortRecursive(arr, pi + 1, high);
    }
}
void quickSort(array *a) {
    quickSortRecursive(a-> arr, 0, a-> size-1);
    return;
}

void radixSort(array *a) {
    fprintf(stdout, "radixSort currently not implemented.\n");
    return;
}

void bucketSort(array *a) {
    fprintf(stdout, "bucketSort currently not implemented.\n");
    return;
}

File: SDAC/Esercitazioni/Esercitazione 4/java-source/Sort.java
import java.util.*;

public class Sort {

    /* Disponibile in libreria java */
    public void quickSortDefault(int[] array) {
        Arrays.sort(array);
    }
    public void swap(int[] array, int first, int last) {
        int temp = array[first];
        array[first] = array[last];
        array[last] = temp;
    }
    public void printArray(int[] array, int n) {
        if(n <= 0) {
            System.out.println("");
            return;
        }
        System.out.print(array[n-1] + " ");
        printArray(array, n-1);
    }

    public void mergeSort(int[] array) {
        //implementazione solo del metodo ricorsivo
		mergeSortAux(array, 0, array.length -1);
        return;
    }
	public void mergeSortAux(int[] array, int left, int right) {
		if(left < right) {
			int mid = (left+right) / 2;
			mergeSortAux(array, left, mid);
			mergeSortAux(array, mid+1, right);
			merge(array, left, mid, right);
		}
	}
	public void merge(int[] array, int left, int mid, int right) {
		int i, j, k;
		int n1 = mid-left+1, n2 = right-mid;
		int L[] = new int[n1];	int R[] = new int[n2]; //creazione di array temporanei

		for(i = 0; i < n1; ++i) {
			L[i] = array[left+i];	//copia dei primi n1 elementi nell'array di sinistra
		}
		for(j = 0; j < n2; ++j) {
			R[j] = array[mid+j+1];	//copia dei primi n2 elementi nell'array di destra
		}

		i = 0;	j = 0;	k = left;

		while(i < n1 && j < n2) {
			if(L[i] < R[j]) {
				array[k] = L[i];
				i++;
			} else {
				array[k] = R[j];
				j++;
			}
			k++;
		}

		while(i < n1) {
			array[k] = L[i];
			i++;
			k++;
		}
		while(j < n2) {
			array[k] = R[j];
			j++;
			k++;
		}
	}

    public void heapSort(int[] array) {
        int n = array.length;
        //Fase 1: Organizzare l'array in un heap
        for(int i = n/2 -1; i >= 0; --i) {
            heapify(array, n, i);
        }

        //Fase 2: Estrarre uno ad uno gli elementi dall'heap
        for(int i = n-1; i > 0; --i) {
            swap(array, 0, i);
            heapify(array, i, 0);
        }
        return;
    }
    public void heapify(int[] array, int n, int i) {
        int largest = i;
        int left = i*2 +1;  //Figlio sinistro
        int right = i*2 + 2;//Figlio destro

        if(left<n && array[left] > array[largest])
            largest = left;
        if(right<n && array[right] > array[largest])
            largest = right;
        
        if(largest != i) {
            swap(array, i, largest);

            heapify(array, n, largest);
        }
    }

    public void insertionSort(int[] array) {
        for(int i = 1; i < array.length; ++i) {
            int key = array[i];
            int j = i-1;

            while(j >= 0 && array[j] > key) {
                array[j+1] = array[j];
                j--;
            }
            array[j+1] = key;
        }
        return;
    }
    public void insertionSortAux(int[] array, int n) {
        //caso base se ha un solo elemento (è già ordinato)
        if(n <= 1) return;

        //ordina i primi n-1 elementi
        insertionSortAux(array, n-1);

        int key = array[n-1];
        insertRicorsivo(array, n-2, key);
    }
    public void insertRicorsivo(int[] array, int j, int key) {
        if(j < 0 || array[j] <= key) { //trovata la posizione corretta o siamo arrivati all'inizio
            array[j+1] = key;
            return;
        }

        //sposto l'elemento a destra e continuo la ricerca
        array[j+1] = array[j];

        //insert ricorsivo chiamato al passo precedente
        insertRicorsivo(array, j-1, key);
    }
    public void insertionSortRicorsivo(int[] array) {
        insertionSortAux(array, array.length);
        return;
    }

    public void selectionSort(int[] array) {
        System.out.println("selectionSort non è ancora implementato");
        return;
    }

    public void quickSort(int[] array) {
        int clone[] = array.clone();
        int r = array.length - 1;
        quickAux(array, 0, r);
        return;
    }
    public void quickAux(int[] array, int p, int r) {
        if(p < r) {
            int q = quickPartition(array, p, r);
            quickAux(array, p, q-1);
            quickAux(array, q+1, r);
        }
    }
    public int quickPartition(int[] array, int p, int r) {
        int x = array[r];
        int i = p-1;
        for(int j = p; j < r; ++j) {
            if(array[j] <= x) {
                i++;
                swap(array, i, j);
            }
        }
        swap(array, i+1, r);
        System.out.print("Array partizionato: ");
        printArray(array, array.length);
        return i+1;
    }

    public void radixSort(int[] array) {
        System.out.println("radixSort non è ancora implementato");
        return;
    }

    public void bucketSort(int[] array) {
        System.out.println("bucketSort non è ancora implementato");
        return;
    }
}
